### PR TITLE
Token authentication

### PR DESCRIPTION
**Pull request by fadezak100**
- Completed VerifyAccessToken Middleware
- Completed VerifyToken controller
- Completed (signToken, generateAccessToken, and verifyToken) Auth helpers functions.
- Completed Tests for both the controller and the auth helpers functions.
- Installed jsonwebtoken, bcrypt *not pretty sure why* and their types npm modules.
- Added '--runInBand' flag to jest to run tests synchronously.
- Added  '--no-cache' flag, not sure why but it solved a problem where jest fails to exist after testing some of the supertest routes.

_There's this problem in testing authHelper functions where functions need types specified, I tried everything but I couldn't figure out their types so I casted them to 'any'._

### CODE REVIEW

File: src/controllers/AuthController.ts [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import { Request, Response } from 'express'
+import { verifyToken } from '../helpers/AuthHelper'
+
+const verifyTokenController = async (req: Request, res: Response) => {
```
Comment (AhmedSafi97) [inline src/controllers/AuthController.ts:4]: You need to create a custom interface that extends the `Reqeust` interface where you need to specify the custom proprieties on this specific request for example this request body has a `token` which is a string. Too much hustle but it is worth it.

```js
export interface verifyTokenControllerRequest extends Request {
  body: { token: string }
}
```

the same can be applied to the response :thinking: ?

Also drop the `Controller` part in the function name, we already can tell what is that from the file name.

File: src/controllers/AuthController.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+  const { token } = req.body
+  try {
+    await verifyToken(token)
+    res.status(200).send({ msg: 'authorized' })
```
Comment (AhmedSafi97) [inline src/controllers/AuthController.ts:8]: plz don't use such a messy shortcuts, just name it message. One more thing it should be `{ 'message' : 'success', 'user': {...} }` here and `{ 'error': 'unauthorized'}`.

File: src/middlewares/AuthMiddleware.ts [hunk 3]
```diff
@@ -1 +1,19 @@
...
+const verifyAccessToken = (req: Request, res: Response, next: NextFunction) => {
+  const { token } = req.body
+
+  if (!token) {
+    throw new Error()
+  }
+
+  try {
+    verifyToken(token)
+    next()
```
Comment (AhmedSafi97) [inline src/middlewares/AuthMiddleware.ts:13]: You need to pass the user id coming from the token through res.locals

File: src/routes/AuthRoute.ts [hunk 4]
```diff
@@ -0,0 +1,8 @@
+import express from 'express'
+import { verifyTokenController } from '../controllers/AuthController'
```
Comment (AhmedSafi97) [inline src/routes/AuthRoute.ts:2]: nah, we will follow a different pattern. call it here AuthController and below `AuthController.verifyToken`

**Review by AhmedSafi97 — CHANGES_REQUESTED** [review]

File: src/routes/AuthRoute.ts (hunk 4 above)
Comment (fadezak100) [reply src/routes/AuthRoute.ts:2]: This makes sense! See, the reason I added `controller` to `verifyToknControllar` is that I had another `verifyToken` function. Kinda mixed names situation.

File: src/controllers/AuthController.ts (hunk 1 above)
Comment (fadezak100) [reply src/controllers/AuthController.ts:4]: I tried doing the same with the response interface. It doesn't work. I may have missed something tho.
Also, I still don't get why returning a custom interface won't work with the `AuthHelpers` functions. I still have to cast them to `any` which doesn't look right!

**Review by AhmedSafi97 — CHANGES_REQUESTED** [review]

**Comment by AhmedSafi97** [comment]
we still need a couple more changes on this PR, declining to make the fixes on a new clean one

**Pull request closed without merging**